Add additional repo card utils from `modelcards` repo

nateraw opened the pull request:
This PR aims to add repo card utilities built out in [nateraw/modelcards](https://github.com/nateraw/modelcards) to this repo. Note that there are some lingering [issues](https://github.com/nateraw/modelcards/issues) we may want to resolve and update both there and here before merging.

Some todos:

- [x] Update existing repo card utils to use new dataclasses + ensure backwards compatibility
- [x] Add default model card and include it in package here
- [x] Do not merge until default model card is updated to latest spec we defined offline
- [x] Add tests from `modelcards` to this repo
  - [x] Add additional tests for private repos
  - [x] (maybe?) Add additional tests that support alternative base endpoint (i.e. staging env)
- [x] Check on min/max versions of Jinja2 and update accordingly in `setup.py`
- [x] Check all docstrings work as intended (original repo tested with doctests, I don't think we do that here?)
- [x] Add section to docs that explains usage

Latest walkthrough of this PR in Colab: [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/gist/nateraw/c5ba361def8fa7c949af851ae2acc7e8/modelcard-in-huggingface-hub-demo.ipynb)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_940). All of your documentation changes will be reflected on that endpoint.

nateraw commented:
This PR is ready for a first round of reviews. The docstrings may be out of date a bit (which is why I'll leave this as draft), but the tests are all in. The example colab in my message above shows the usage. 

I don't want to add anything in docs for now in case there are big changes from review I need to make.

LysandreJik commented on src/huggingface_hub/repocard.py:
```diff
@@ -1,30 +1,288 @@
+        Args:
+            filepath (`Union[Path, str]`): Filepath to the markdown file to save.
+
+        Example:
+            >>> from huggingface_hub import RepoCard
+            >>> card = RepoCard("---\nlanguage: en\n---\n# This is a test repo card")
+            >>> card.save("/tmp/test.md")
```
```suggestion
        Example:
            ```python
            >>> from huggingface_hub import RepoCard
            >>> card = RepoCard("---\nlanguage: en\n---\n# This is a test repo card")
            >>> card.save("/tmp/test.md")
            ```
```

LysandreJik replied:
(please apply the same to all code examples)

nateraw replied:
is it ok if I add a new line before the closing backticks? this way doctests still run as expected (even though they aren't being used in our CI)

LysandreJik commented on src/huggingface_hub/repocard.py:
```diff
@@ -1,30 +1,288 @@
+            token (`str`, *optional*):
+                Authentication token, obtained with `huggingface_hub.HfApi.login` method. Will default to
+                the stored token.
+            repo_type (`str`, *optional*):
```
```suggestion
            repo_type (`str`, *optional*, defaults to `"model"`):
```

LysandreJik commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,369 @@
+
+        Args:
+            language (`Union[str, List[str]]`, *optional*):
+                Language of model's training data or metadata. Example: `'en'`. Defaults to `None`.
```
Maybe specify these are `ISO 639-1` language codes

julien-c replied:
technically the backend validation says this:

> It must be an ISO 639-1, 639-2 or 639-3 code (two/three letters), or a special value like "code", "multilingual"

LysandreJik commented on tests/test_modelcard.py:
```diff
@@ -0,0 +1,199 @@
+import logging
```
Please align with other tests in the library by putting all test in a class that inherits from `unittest.TestCase`, and using `self.assertXXX` instead of `assert` statements.

nateraw replied:
:( It looks so much cleaner this way though, imo. But okay 😛

LysandreJik replied:
Cleanliness is a subjective point of view while conventions are not :wink:

adrinjalali commented on setup.py:
```diff
@@ -19,6 +19,7 @@ def get_version() -> str:
     "typing-extensions>=3.7.4.3",  # to be able to import TypeAlias
     "importlib_metadata;python_version<'3.8'",
     "packaging>=20.9",
+    "Jinja2",
```
I'd say jinja templates are what we should use for modelcards:
- Users can define their own template and pass it to these tools, and it's a well documented format which can easily be extended.
- we're adding html and js code in the modelcard, and HTML is not a [regular language](https://en.wikipedia.org/wiki/Regular_language) and therefore [cannot be parsed with a regex](https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454)

But I completely agree that jinja is probably not the best dependency for hfh, and that's partly why I've advocated for modelcards to be its own project/repo rather than being inside hfh.

osanseviero replied:
Friendly ping on this to reach a final consensus on this

julien-c replied:
let's add it as optional dependency i.e. only install it when doing it `pip install -e .[cards]` or equivalent?

LysandreJik replied:
Yes, if adding jinja2 to the `extras_require` is an option, then I'd be more than happy to have this merged in. This extra requirement can be required by the downstream libraries that install `huggingface_hub` in order to manage their model cards, so it's a good compromise for me.

julien-c reviewed (commented):
Played with it a little bit, I really like it! 

This looks to be in good shape.

Aside from small comments, I wonder why you didn't keep the notion of `ModelCard` inheriting from `RepoCard` (which applies to all repo types) which i quite liked in https://github.com/nateraw/modelcards ?
And same for `CardData`, IMO we should have a repo-type-agnostic implem of `CardData` and then children dataclasses that define the relevant fields

My other question was, do we really want to bundle the (or a?) model card template in the library, or should we host it on the Hub and load the template from there too. 
I think there might be advantages and drawbacks to both approaches.

nateraw commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,369 @@
+    verified: Optional[bool] = None
+
+
+@dataclass
+class CardData:
```
I did give that a go here the other day: https://github.com/nateraw/modelcards/pull/57

nateraw replied:
I believe separating them all out makes it harder to read, but "better" structurally. I think its a little too complicated for what's really going on, but if you folks prefer going that direction, I have no problem updating to do that here 😄

osanseviero commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,369 @@
+class CardData:
+    def __init__(
+        self,
+        language: Optional[Union[str, List[str]]] = None,
```
WDYT of forcing users to pass all the parameters by keyword. It will make it easier to keep this up to date and there will be less issues with wrong order of params. E.g. a la https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/hf_api.py#L1428-L1438

Wauplin replied:
👍 for this

nateraw replied:
yea probably a good call! I only intended users to use kwargs anyways, so this makes sense to me.

osanseviero commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,369 @@
+            if type(self.eval_results) == EvalResult:
+                self.eval_results = [self.eval_results]
+            if self.model_name is None:
+                raise ValueError("`eval_results` requires `model_name` to be set.")
```
```suggestion
                raise ValueError("Passing `eval_results` requires `model_name` to be set.")
```

osanseviero commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,369 @@
+    """Takes in a model index and returns a list of `huggingface_hub.EvalResult` objects.
+
+    A detailed spec of the model index can be found here:
+    https://github.com/huggingface/hub-docs/blob/main/modelcard.md
```
Maybe https://github.com/huggingface/hub-docs/blob/main/modelcard.md?plain=1 is a better link

osanseviero commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,369 @@
+        Hugging Face Hub.
+
+    Returns:
+        - model_name (`str`):
```
I'm not sure this format works with the doc builder

osanseviero commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,369 @@
+
+
+def model_index_to_eval_results(model_index: List[Dict[str, Any]]):
+    """Takes in a model index and returns a list of `huggingface_hub.EvalResult` objects.
```
This is actually returning a tuple of a string and a list iiuc

osanseviero commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,369 @@
+        ...         "name": "my-cool-model",
+        ...         "results": [
+        ...         {
+        ...             "task": {
```
Nit: the formatting is a bit broken across this. Some indents are missing

nateraw replied:
all the docstrings are a bit screwed up atm.

osanseviero commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,369 @@
+        ...         ]
+        ...     }
+        ... ]
+        >>> model_name, eval_results = model_index_to_eval_results(model_index)
```
Do we really need to return the `model_name`? I would have expected this to return the `eval_results` only.

nateraw replied:
Model name is required in the eval results.

osanseviero commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,369 @@
+
+    # Metrics are reported on a unique task-and-dataset basis.
+    # Here, we make a map of those pairs and the associated EvalResults.
+    task_and_ds_types_map = dict()
+    for eval_result in eval_results:
+        task_and_ds_pair = (eval_result.task_type, eval_result.dataset_type)
+        if task_and_ds_pair in task_and_ds_types_map:
+            task_and_ds_types_map[task_and_ds_pair].append(eval_result)
+        else:
+            task_and_ds_types_map[task_and_ds_pair] = [eval_result]
```
Alternatively

```suggestion
    task_and_ds_types_map = defaultdict(list)
    for eval_result in eval_results:
        task_and_ds_pair = (eval_result.task_type, eval_result.dataset_type)
        ask_and_ds_types_map[task_and_ds_pair].append(eval_result)
```

nateraw replied:
hmm will give this one a go, should work.

osanseviero commented on src/huggingface_hub/repocard_types.py:
```diff
@@ -1,105 +0,0 @@
-from dataclasses import dataclass, field
```
Bye bye code :wave:

nateraw replied:
hehe

osanseviero reviewed (commented):
Thanks for this PR! It's very cool! :fire: I left some questions and high-level suggestions, but this is in a great state already!

* First of all, thanks for sharing the colab! It makes it super easy and nice to play with the API very quickly. I think a lot of it can be converted to some nice docs :mechanical_arm: 
* I agree with @julien-c on having different classes for each type of repo card, with `ModelCard`, `SpacesCard`, etc.
* I think having the main template bundled in the library is nice, so I would keep as is. If people want to provide new templated in the future we could have some additional functionality to use templates from the Hub, which might motivate people to create their own templates, but I think for now focusing in the current one is better!
* I didn't review the tests yet as they are probably going to change a bit :) 

Thanks again! :rocket:

nateraw commented on setup.py:
```diff
@@ -19,6 +19,7 @@ def get_version() -> str:
     "typing-extensions>=3.7.4.3",  # to be able to import TypeAlias
     "importlib_metadata;python_version<'3.8'",
     "packaging>=20.9",
+    "Jinja2",
```
Opted to just put it in `testing` requirements. Since Jinja is only required by `from_template` and its only one dep, I figure its a bit overkill to add a whole new extras category.

osanseviero replied:
Should this be removed from `install_requires` ?

nateraw replied:
Whoops! you're right.

nateraw commented on src/huggingface_hub/repocard.py:
```diff
@@ -1,30 +1,288 @@
+
+        return cls(Path(card_path).read_text())
+
+    def validate(self, repo_type="model"):
```
I actually find it fairly useful. Also, by leaving it public we can document it in reference docs so users understand what's going on if something fails.

nateraw commented on src/huggingface_hub/repocard.py:
```diff
@@ -1,30 +1,288 @@
+    def validate(self, repo_type="model"):
+        """Validates card against Hugging Face Hub's model card validation logic.
+        Using this function requires access to the internet, so it is only called
+        internally by `huggingface_hub.ModelCard.push_to_hub`.
```
Had this discussion w/ Adrin in original PR in `modelcards`, we opted to have it not return anything. It errors out if theres an issue, so that's how we're communicating it vs. returning `True`/`False`.

nateraw commented:
Ok folks, I've:

- Updated tests to `unittest.TestCase` instead of `pytest`
- Refactored to have `Repocard` be parent of `ModelCard` and `DatasetCard`
- Refactored to have `CardData` be parent of `ModelCardData` and `DatasetCardData`
- Updated to have Jinja2 not be installed by default, only as extra
- Added docs (both a guide + reference documentation)

osanseviero commented on docs/source/how-to-model-cards.mdx:
```diff
@@ -0,0 +1,307 @@
+
+## Loading a Model Card from the Hub
+
+To load an existing card from the hub, you can use the [`ModelCard.load`] function. Here, we'll load the card from [`nateraw/vit-base-beans`](https://huggingface.co/nateraw/vit-base-beans).
```
```suggestion
To load an existing card from the Hub, you can use the [`ModelCard.load`] function. Here, we'll load the card from [`nateraw/vit-base-beans`](https://huggingface.co/nateraw/vit-base-beans).
```

osanseviero commented on docs/source/how-to-model-cards.mdx:
```diff
@@ -0,0 +1,307 @@
+
+Another way you might want to do this is with f-strings. In the following example, we:
+
+- Use [`ModelCardData.to_yaml`] to convert metadata we defined to YAML so we can use it to insert the YAML block in the model card
```
```suggestion
- Use [`ModelCardData.to_yaml`] to convert metadata we defined to YAML so we can use it to insert the YAML block in the model card.
```

osanseviero replied:
`ModelCardData` has not been mentioned so far, so maybe adding one sentence about it before could be useful

osanseviero commented on docs/source/how-to-model-cards.mdx:
````diff
@@ -0,0 +1,307 @@
+This model was created by [@nateraw](https://hf.co/nateraw).
+```
+
+As you update the card data, you can validate the card is still valid against the Hub by calling [`ModelCard.validate`]. This ensures that the card passes any validation rules set up on the Hugging Face Hub.
````
Maybe it's worth adding some links to jinja documentation for people that want to create their templates.

nateraw replied:
Done!

osanseviero commented on docs/source/how-to-model-cards.mdx:
```diff
@@ -0,0 +1,307 @@
+
+## Sharing Model Cards
+
+If you're authenticated with the Hugging Face Hub (either by using `huggingface-cli login` or `huggingface_hub.notebook_login()`), you can push cards to hub by simply calling [`ModelCard.push_to_hub`]. Let's take a look at how to do that...
```
```suggestion
If you're authenticated with the Hugging Face Hub (either by using `huggingface-cli login` or `huggingface_hub.notebook_login()`), you can push cards to the Hub by simply calling [`ModelCard.push_to_hub`]. Let's take a look at how to do that...
```

osanseviero commented on docs/source/how-to-model-cards.mdx:
````diff
@@ -0,0 +1,307 @@
+card.push_to_hub(repo_id, create_pr=True)
+```
+
+A resulting PR created from this command can be seen [here](https://huggingface.co/nateraw/hf-hub-modelcards-pr-test/discussions/3).
````
Very cool!

osanseviero commented on docs/source/how-to-model-cards.mdx:
```diff
@@ -0,0 +1,307 @@
+
+### Including Evaluation Results
+
+To include evaluation results in the metadata `model-index`, you can pass an [`EvalResult`] or a list of `EvalResult` with your associated evaluation results. Under the hood it'll create the `model-index` when you call `card.data.to_dict()`.
```
Maybe link to https://huggingface.co/docs/hub/models-cards#evaluation-results

nateraw replied:
Done!

osanseviero commented on src/huggingface_hub/file_download.py:
```diff
@@ -151,6 +159,14 @@ def get_fastcore_version():
+    return _jinja_available
+
+
+def get_jinja_version():
```
I don't think this is used anywhere. Should we remove it?

osanseviero replied:
You're right! Yes, let's do this but in a separate PR as it came before.

osanseviero commented on tests/test_repocard.py:
```diff
@@ -360,3 +388,346 @@ def test_update_new_result_new_dataset(self):
+        self.assertIsNone(card.data.eval_results)
+
+    def test_load_model_card_from_file(self):
+
```
```suggestion
```

osanseviero commented on tests/test_repocard_data.py:
```diff
@@ -0,0 +1,197 @@
+        ]
+        model_name, eval_results = model_index_to_eval_results(model_index)
+
+        self.assertEqual(len(eval_results), 3)
```
hmmm I Might be missing something but aren't there only 2 eval results above?

osanseviero replied:
Ah we have two metrics in the first result. Shouldn't there be 2 metrics in the same result or do we split into multiple results?

nateraw replied:
There are 3 eval results:
- 2 for the cats-vs-dogs dataset and 
- one for the beans dataset. 

The `EvalResult` object is a flattened representation, so each instance represents one evaluation result coming from one dataset.

nateraw replied:
So, the above would be something like this:

```python
from huggingface_hub import EvalResult

eval_results = [
    EvalResult(dataset_type='cats_vs_dogs', metric_type='acc', metric_value=0.85, ...),
    EvalResult(dataset_type='cats_vs_dogs', metric_type='f1', metric_value=0.9, ...),
    EvalResult(dataset_type='beans', metric_type='acc', metric_value=0.9, ...)
]
```

osanseviero replied:
Makes sense, thanks for explaining!

osanseviero reviewed (approved):
Very cool PR! This is very very nice, and the documentation for it is very helpful!

Wauplin commented on tests/test_repocard.py:
```diff
@@ -119,7 +134,20 @@
+    Decorator marking a test that requires Jinja2.
+
+    These tests are skipped when Jinja2 is not installed.
+
```
```suggestion
```

Wauplin reviewed (commented):
Just finished to review this PR. Thanks A LOT @nateraw for all the impressive implementation, documentation and testing. I have left a bunch of small comments but it's overall very good work 👍 🔥 🔥

osanseviero commented on src/huggingface_hub/repocard.py:
```diff
@@ -1,30 +1,288 @@
+        return url
+
+    @classmethod
+    def from_template(
```
No strong opinion to be honest. `from_data` makes sense to me. WDYT @nateraw ?

julien-c replied:
kinda like `from_data` too

nateraw replied:
I'm not a big fan of `from_data`, as I think it's a bit more misleading. We aren't really creating the card 100% from data - a lot of the content is going to come from the template. And keeping the name with "template" in it at least makes sense when you get an error about not having Jinja2 (imo). 

I could be convinced to change it, though, if anybody is very opinionated about it!

Wauplin replied:
Seems no one is so opinionated but we still don't have a consensus though 😁. 
Maybe `RepoCard.render_with_data` (or `RepoCard.render_from_data(...)`) is more convincing ? It's less the name that I would give to a classmethod but we have both the templating and the data aspects.

Wauplin replied:
And in any case, if no naming reach a perfect consensus I suggest with keep `RepoCard.from_template` just because it's good enough and so that we can close this (huge) PR :)

adrinjalali replied:
As an alternative design, the pattern we chose in `skops` is for the user to create an instance of the object, then add contents for the sections through instance methods. From the users' perspective, the code looks like this, and the implementation is here: https://github.com/skops-dev/skops/blob/main/skops/card/_model_card.py

``` python
# this constructor would take a path to a template in a future iteration, for now supporting only
# the default template.
# Instead of calling the metadata `CardData`, we call it `metadata` which is easier
# to understand from the users' perspective.
model_card = card.Card(model=sklearn_estimator, metadata=...)

# The user can also access and change/add metadata by accessing the metadata object
model_card.metadata.license = "mit"

limitations = "This model is not ready to be used in production."
model_description = (
    "This is a HistGradientBoostingClassifier model trained on breast cancer dataset."
    " It's trained with Halving Grid Search Cross Validation, with parameter grids on"
    " max_leaf_nodes and max_depth."
)
model_card_authors = "skops_user"
get_started_code = (
    "import pickle\nwith open(pkl_filename, 'rb') as file:\n    clf = pickle.load(file)"
)
citation_bibtex = "bibtex\n@inproceedings{...,year={2020}}"

# they can add info in a single call, or chain calls to `add`, since all `add*` methods
# return self.
model_card.add(
    citation_bibtex=citation_bibtex,
    get_started_code=get_started_code,
    model_card_authors=model_card_authors,
    limitations=limitations,
    model_description=model_description,
)

# adding metrics
eval_descr = (
    "The model is evaluated on test data using accuracy and F1-score with macro"
    " average."
)
model_card.add(eval_method=eval_descr)

...
model_card.add_metrics(**{"accuracy": accuracy, "f1 score": f1})

# or save and add plots
...
model_card.add_plot(**{"Confusion matrix": "confusion_matrix.png"})

# or add tabular information about hyper parameter search space and resuts
...
model_card.add_table(
    # they decide if the table should be in a `<details>` tag.
    folded=True,
    **{
        "Hyperparameter search results": cv_results,
        "Classification report": clf_report,
    },
)

# And finally, this is how the model is _rendered_ into a file.
model_card.save(Path(local_repo) / "README.md")

# However, they could also render the model into a string:
rendered = model_card.render()
```

nateraw replied:
I prefer that we leave it as `from_template` for now - I am opinionated enough to push for that. I think it makes it very clear that we're using a 'template', which implies the need for the extra `Jinja2` dep. I promise I'm not just being lazy and not wanting to change the name 😅 .

As for the usage described by @adrinjalali - I prefer the way we have implemented here, as it results in less code I think, and effectively does the same thing.

nateraw commented on src/huggingface_hub/templates/datasetcard_template.md:
```diff
@@ -0,0 +1,126 @@
+
+### Dataset Summary
+
+[More Information Needed]
```
Yes we should definitely make it more interactive. Perhaps I'll leave that to separate, smaller PR since this one has gotten so big

nateraw commented on src/huggingface_hub/repocard.py:
```diff
@@ -131,78 +571,75 @@ def metadata_eval_result(
+        ...         }
+        ...     ]
+        ... }
+        True
```
this is a doctest, so `results == ...` is just saying the results I got on the previous line equal the dict you're seeing. The `True` is the output. We arent running doctests in CI, but they are really useful to make sure the example snippets actually work.

Wauplin replied:
Ok thanks for the explanation. Never used it but it seems cool 👍

nateraw commented on src/huggingface_hub/repocard.py:
```diff
@@ -1,30 +1,470 @@
+            )
+
+        template_path = template_path or cls.default_template_path
+        content = jinja2.Template(Path(template_path).read_text()).render(
+            card_data=card_data.to_yaml(), **template_kwargs
+        )
```
Sure, I dig this...I think I actually had it written similar to this in one of the earlier versions of `modelcards`.

Anything to think about wrt whether this could break something/cause some unexpected behavior?

Wauplin replied:
As long as we create a new dictionary with `.copy()` and that we give priority to user-defined kwargs I don't see what could go wrong.

A problem can happen if a user defines its own jinja template, use an optional variable in it that has the same name as an attribute of `RepoCardData` -without having the same value/same meaning- and that we overwrite it from card_data even though the user did not pass it. Said like this it seems a quite specific use-case and the user always have the possibility to update its template to use a different var name.

Always difficult to look for a blind spot though 😫

nateraw commented on src/huggingface_hub/repocard.py:
```diff
@@ -1,30 +1,470 @@
+    @classmethod
+    def from_template(
+        cls,
+        card_data: CardData,
```
v nice catch, thanks

nateraw commented on src/huggingface_hub/repocard.py:
```diff
@@ -1,30 +1,470 @@
+
+
+def _detect_line_ending(content: str) -> Literal["\r", "\n", "\r\n", None]:
+    """
+    same implem as in Hub server, keep it in sync
+    """
```
@julien-c do you have a good description for this? 😅 I didn't really fully understand when you issued the pr for it, so maybe you'd be best to explain

Wauplin replied:
Then a description is definitely needed 😁

julien-c replied:
hmm yeah it's just a helper to determine the line endings for a file

This is useful when updating a model card, e.g. changing just one metadata item, on a model that was created in Windows. If you don't honor the existing line ending you get a diff on all lines (and the resulting file has mixed line ending)

julien-c replied:
not a huge deal considering most users have \n line endings, but we encountered this issue when doing mass-modification of model cards on the Hub (affected a few % of the model cards)

nateraw commented on src/huggingface_hub/repocard.py:
```diff
@@ -265,149 +690,91 @@ def metadata_update(
+        ...                                        'value': 0.7762102282047272}],
+        ...                          'task': {'name': 'Text Classification',
+        ...                                   'type': 'text-classification'}}]}]}
+        >>> url = metadata_update("julien-c/reactiongif-roberta", metadata)
```
TODO - use internal testing repo instead. This one keeps getting blown up whenever I run doctests 😂

nateraw commented on src/huggingface_hub/repocard.py:
```diff
@@ -1,30 +1,470 @@
+            r.raise_for_status()
+        except requests.exceptions.HTTPError as exc:
+            if r.status_code == 400:
+                raise RuntimeError(r.text)
```
SGTM, updated.

nateraw commented:
Ok folks, I'm ready to call this good to go. Any small updates to docs/etc can be done in future small PRs, as long as there's nothing major lingering that anybody is super opinionated about.

Feel free to give this a final look over.

---

@Wauplin, as a sanity check, can you please double check I didn't mess anything up when I rebased? one of my most [recent commits](https://github.com/huggingface/huggingface_hub/pull/940/commits/8ea02af2feb619c730d364c472054c5fab178ca9) ended up showing me codecov twice, causing tests to fail for some reason, but it resolved itself when I pushed to this branch again.

Should be fine, just would like another pair of eyes on it. Thank you!!

meg-huggingface reviewed (commented):
Great work!!

julien-c commented on docs/source/package_reference/cards.mdx:
```diff
@@ -0,0 +1,40 @@
+[[autodoc]] ModelCard
+
+[[autodoc]] ModelCardData
+
```
i'm not too bent on the name but on HF side i like that symmetry:

`A X Card is materialized by a README.md file inside a X repo` where X can be {model, dataset, Space}

julien-c replied:
we can always document that it's "also usually known as a Data Card", ofc

Wauplin replied:
I added a line in @nateraw's documentation for that

Wauplin commented on docs/source/how-to-model-cards.mdx:
```diff
@@ -0,0 +1,317 @@
+    model_description="this model does this and that",
+    developers="Nate Raw",
+    more_resources="https://github.com/huggingface/huggingface_hub",
+    **card_data.to_dict()  # Pass along any card data vals that might fill out part of the template
```
```suggestion
```

No need for this anymore as card_data values are passed along by default

Wauplin commented on docs/source/how-to-model-cards.mdx:
```diff
@@ -0,0 +1,317 @@
+    model_description="this model does this and that",
+    developers="Nate Raw",
+    more_resources="https://github.com/huggingface/huggingface_hub",
+    **card_data.to_dict()  # Pass along any card data vals that might fill out part of the template
```
```suggestion
```

same here

Wauplin commented on docs/source/package_reference/cards.mdx:
```diff
@@ -0,0 +1,40 @@
+
+The `RepoCard` object is the parent class of [`ModelCard`] and [`DatasetCard`].
+
+[[autodoc]] huggingface_hub.repocard.RepoCard
+
```
```suggestion
[[autodoc]] huggingface_hub.repocard.RepoCard
    - __init__
    - all
```

Repocard init does a lot of things and is really well documented so I would add it explicitely to the generated doc.

Wauplin commented on docs/source/package_reference/cards.mdx:
```diff
@@ -0,0 +1,40 @@
+
+[[autodoc]] huggingface_hub.repocard.RepoCard
+
+The [`CardData`] object is the parent class of [`ModelCardData`] and [`DatasetCardData`].
```
```suggestion
## Card Data

The [`CardData`] object is the parent class of [`ModelCardData`] and [`DatasetCardData`].
```

Wauplin commented on docs/source/package_reference/cards.mdx:
```diff
@@ -0,0 +1,40 @@
+[[autodoc]] huggingface_hub.repocard_data.CardData
+
+## Model Cards
+
+[[autodoc]] ModelCard
+
+[[autodoc]] ModelCardData
```
```suggestion
### ModelCard

[[autodoc]] ModelCard

### ModelCardData

[[autodoc]] ModelCardData
```

Wauplin commented on docs/source/package_reference/cards.mdx:
```diff
@@ -0,0 +1,40 @@
+
+[[autodoc]] ModelCardData
+
+## Dataset Cards
+
+[[autodoc]] DatasetCard
+
+[[autodoc]] DatasetCardData
+
```
```suggestion
## Dataset Cards

Dataset cards are also known as Data Cards in the ML Community.

### DatasetCard

[[autodoc]] DatasetCard

### DatasetCardData

[[autodoc]] DatasetCardData

```

Wauplin commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,538 @@
+        dataset_revision (`str`, *optional*):
+            The revision (AKA Git Sha) of the dataset used in `load_dataset()`.
+            Example: 5503434ddd753f426f4b38109466949a1217c2bb
+        dataset_args (`Dict[str, Any]`, *optional*):
+            The arguments passed during `Metric.compute()`. Example for `bleu`: max_order: 4.
```
```suggestion
        dataset_args (`Dict[str, Any]`, *optional*):
            The arguments passed during `Metric.compute()`. Example for `bleu`: `{"max_order": 4}`
```

Wauplin commented on docs/source/package_reference/cards.mdx:
```diff
@@ -0,0 +1,40 @@
+
+[[autodoc]] DatasetCardData
+
+## Utilities
+
+[[autodoc]] EvalResult
+
+[[autodoc]] huggingface_hub.repocard_data.model_index_to_eval_results
+
+[[autodoc]] huggingface_hub.repocard_data.eval_results_to_model_index
+
+[[autodoc]] huggingface_hub.repocard.metadata_eval_result
+
+[[autodoc]] huggingface_hub.repocard.metadata_update
```
```suggestion
## Utilities

### EvalResult

[[autodoc]] EvalResult

### model_index_to_eval_results

[[autodoc]] huggingface_hub.repocard_data.model_index_to_eval_results

### eval_results_to_model_index

[[autodoc]] huggingface_hub.repocard_data.eval_results_to_model_index

### metadata_eval_result

[[autodoc]] huggingface_hub.repocard.metadata_eval_result

### metadata_update

[[autodoc]] huggingface_hub.repocard.metadata_update
```

Wauplin commented on src/huggingface_hub/repocard_data.py:
```diff
@@ -0,0 +1,538 @@
+            Example: 5503434ddd753f426f4b38109466949a1217c2bb
+        dataset_args (`Dict[str, Any]`, *optional*):
+            The arguments passed during `Metric.compute()`. Example for `bleu`: `{"max_order": 4}`
+        metric_name: (`str`, *optional*):
+            A pretty name for the metric. Example: "Test WER".
+        metric_config: (`str`, *optional*):
+            The name of the metric configuration used in `load_metric()`.
+            Example: bleurt-large-512 in `load_metric("bleurt", "bleurt-large-512")`.
+            See the `datasets` docs for more info: https://huggingface.co/docs/datasets/v2.1.0/en/loading#load-configurations
+        metric_args: (`Dict[str, Any]`, *optional*):
+            The arguments passed during `Metric.compute()`. Example for `bleu`: max_order: 4
+        verified: (`bool`, *optional*):
```
```suggestion
        metric_name (`str`, *optional*):
            A pretty name for the metric. Example: "Test WER".
        metric_config (`str`, *optional*):
            The name of the metric configuration used in `load_metric()`.
            Example: bleurt-large-512 in `load_metric("bleurt", "bleurt-large-512")`.
            See the `datasets` docs for more info: https://huggingface.co/docs/datasets/v2.1.0/en/loading#load-configurations
        metric_args (`Dict[str, Any]`, *optional*):
            The arguments passed during `Metric.compute()`. Example for `bleu`: max_order: 4
        verified (`bool`, *optional*):
```

Wauplin commented:
@nateraw I double-checked everything and I think we are good to go.
Thanks again for the huge work done in this PR both in the implementation and in documenting everything !
